### PR TITLE
Fix group_by with 0 groups

### LINKED ISSUE #3935: group_by() on 0 rows tibble
Is this the right behavior: 

``` r
library(dplyr)

tibble(x = integer(), y = character()) %>% 
  group_by(y) %>% 
  group_data()
#> # A tibble: 1 x 2
#>   y     .rows    
#>   <chr> <list>   
#> 1 <NA>  <int [0]>
```

Created on 2018-10-22 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #3935. 

```
> tibble(x = integer()) %>% group_by(x) %>% group_data()
# A tibble: 0 x 2
# ... with 2 variables: x <int>, .rows <list>
```

### CODE REVIEW

**Comment by romainfrancois** [comment]
Still need some work for summaries, e.g. : 

```r
> tibble(x = integer()) %>% group_by(x) %>% summarise(y = +mean(x)) 
# A tibble: 0 x 2
# ... with 2 variables: x <int>, y <lgl>
```

The type of `y` is wrong. That comes from https://github.com/tidyverse/dplyr/blob/53ed25b65a8db4b619ad7212863c6532b66c8021/inst/include/dplyr/standard/GroupedCallReducer.h#L405

**Comment by hadley** [comment]
I don't think we have any way to the types out of `summarise()` to be totally correct, so I think the best option is just to file an option in vctrs to make sure we consider this problem.

**Comment by romainfrancois** [comment]
```r
> tibble(x = integer()) %>% group_by(x) %>% summarise(y = +mean(x), z = list()) 
# A tibble: 0 x 3
# ... with 3 variables: x <int>, y <dbl>, z <list>
```

I've also updated `nest_by` so that it keeps a "model" of the nested data frame, mainly for the purpose of `unnest()`: 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
df <- tibble(x=integer(), y = character())

df %>% 
  nest_by(y) %>% 
  tidyr::unnest()
#> # A tibble: 0 x 2
#> # ... with 2 variables: y <chr>, x <int>
```

(needs that PR in tidy): https://github.com/tidyverse/tidyr/pull/511

File: R/nest_by.R [hunk 1]
```diff
@@ -2,10 +2,13 @@
 nest_by_impl <- function(.data, key_var){
   to_nest <- select(ungroup(.data), setdiff(tbl_vars(.data), group_vars(.data)))
 
-  .data %>%
+  out <- .data %>%
     group_data() %>%
     mutate(!!key_var := map(.data$.rows, function(.x) to_nest[.x,, drop = FALSE])) %>%
     select(-".rows")
+
+  attr(out[[key_var]], "model") <- to_nest[integer(), , drop = FALSE]
```
Comment (krlmlr) [inline R/nest_by.R]: What if we nest more than once?

Comment (krlmlr) [reply R/nest_by.R]: Never mind, I ignored that this is an attribute of the column, multiple nesting will likely work well.

Do we need to be consistent with `list_of()` at this stage? https://github.com/r-lib/vctrs/blob/796a03bd72c7b75f36eec73b69df1580c832afc0/R/list_of.R#L18

Comment (romainfrancois) [reply R/nest_by.R]: Yeah maybe `list_of` gives the right abstraction for this.

**Pull request merged**